text_generation/causal_lm/cpp/speculative_decoding.cpp

```diff
@@ -0,0 +1,379 @@
+        
+        for (auto name: {"key", "value"}) {
+            ss << "present." << i;
+            ss << (name == "key" ? ".key" : ".value");
+            auto tensor_name = ss.str();
+            
+            auto trimmed_tensor = trimm_tensor(request.get_tensor(tensor_name), seq_len_axis, new_seq_len);
+            request.set_tensor(tensor_name, trimmed_tensor);
+            ss.str("");
+        }
+    }
+}
+
+bool verbose = false;
+
+using namespace std;
+void print_accum_tokens(ov::InferRequest& detokenizer, std::vector<int64_t> tokens, string suffix) {
+    std::string text = detokenize(detokenizer, tokens);
+    stringstream ss;
+    for (const auto& token: tokens)
+        ss << token << " ";
+    cerr << suffix << ": ";
+    cerr << text << " | " << ss.str() << endl;
+}
+
+string tokens_to_string(ov::InferRequest& detokenizer, std::vector<int64_t> tokens) {
+    return detokenize(detokenizer, tokens);
+}
+
+int main(int argc, char* argv[]) try {
+    int draft_kv_size = 22;
+    int target_kv_size = 32;
+
+    // target_kv_size = 22;  // if TinyLlama is used as targed for testing purposes
+
+    if (argc != 4 && argc != 5) {
+        throw std::runtime_error(std::string{"Usage: "} + argv[0] + " <DRAFT MODEL_DIR> <TARGET MODEL_DIR> '<PROMPT>'");
+    }
+    if (argc == 5) {
+        if (!strcmp(argv[4], "verbose"))
+            verbose = true;
+    }
+    // Compile models
+    ov::Core core;
+    core.add_extension(OPENVINO_TOKENIZERS_PATH);  // OPENVINO_TOKENIZERS_PATH is defined in CMakeLists.txt
+    // tokenizer and detokenizer work on CPU only
+    ov::InferRequest tokenizer = core.compile_model(
+        std::string{argv[1]} + "/openvino_tokenizer.xml", "CPU").create_infer_request();
+    auto [input_ids, attention_mask] = tokenize(tokenizer, argv[3]);
+    ov::InferRequest detokenizer = core.compile_model(
+        std::string{argv[1]} + "/openvino_detokenizer.xml", "CPU").create_infer_request();
+    // todo: for debug reason add one more detokenizer since the original one is moved to TextStreamer
+    ov::InferRequest detokenizer_2 = core.compile_model(
+        std::string{argv[1]} + "/openvino_detokenizer.xml", "CPU").create_infer_request();
+    
+    // draft model
+    ov::InferRequest lm = core.compile_model(
+        std::string{argv[1]} + "/openvino_model.xml", "CPU").create_infer_request();
+
+    lm.set_tensor("input_ids", input_ids);
+    // std::fill_n(input_ids.data<int64_t>(), input_ids.get_size(), 1);
+    uint64_t iter = input_ids.get_shape()[1] + 1;
+
+    lm.set_tensor("input_ids", input_ids);
+    lm.set_tensor("attention_mask", attention_mask);
+    ov::Tensor position_ids = lm.get_tensor("position_ids");
+    position_ids.set_shape(input_ids.get_shape());
+    std::iota(position_ids.data<int64_t>(), position_ids.data<int64_t>() + position_ids.get_size(), 0);
+    init_key_values(lm, draft_kv_size);
+    lm.infer();
+
+    // target
+    ov::InferRequest lm_target = core.compile_model(
```

Comment:
Using TinyLlama-1.1B-Chat-v1.0 for both draft and target also results in divergence. I wasn't able to identify the problem, but maybe it will be easier to debug it this way

Reply:
i find where divergence occurs, and fixed. Now speculative sampling and greedy causal return exactly the same values with 1.5x performance increase.